wordpress-mobile/WordPress-FluxC-Android
# Fetch coupons

This PR implements the endpoint for fetching coupons.

Since it involves new functionality, I added the new tables to the Room DB. Coupons can reference products and product categories and there is many-to-many relationship between the entities, so there are also new cross-reference tables. Because I wanted to construct a full model tree when serving the result, I also added a new `Products` and `ProductCategories` tables to Room. When products and categories are added, updated or deleted, the parallel Room tables are also updated.

The `Products` and `ProductCategories` tables are incomplete (I left out the referenced entities, which can be added later on when we migrate the tables fully to Room).

This PR has a lot of line changes because the schema json has 1200+ lines and the table definitions are also very spread out. If you'd like me to split the PR up, let me know.

@wzieba I added you as a reviewer, because you've started adding Room tables, sorry :). 

I'll be adding unit tests in a separate PR.

**To test:**
1. Start the example app 
2. Log in and go to Woo -> Select site (make sure you have some coupons ready)
3. Fetch coupons
4. Notice the coupons are fetched (you can also verify this in the DB inspector)
